table/tables/state_remote.go

```diff
@@ -0,0 +1,277 @@
+type lockForWriteOP struct {
+	sync.WaitGroup
+	// Input
+	tid int64
+	now uint64
+	ts  uint64
+
+	// Output
+	err      error
+	oldLease uint64
+}
+
+func (op *lockForWriteOP) Exec(data *mockStateRemoteData) {
+	op.oldLease, op.err = data.LockForWrite(op.tid, op.now, op.ts)
+	op.Done()
+}
+
+type mockStateRemoteData struct {
+	data map[int64]*stateRecord
+}
+
+type stateRecord struct {
+	lockLease    uint64
+	oldReadLease uint64 // only use for intent lock, it means old read lease.
+	lockType     CachedTableLockType
+}
+
+func newMockStateRemoteData() *mockStateRemoteData {
+	return &mockStateRemoteData{
+		data: make(map[int64]*stateRecord),
+	}
+}
+
+func (r *mockStateRemoteData) Load(tid int64) (CachedTableLockType, uint64, error) {
+	record, ok := r.data[tid]
+	if !ok {
+		return CachedTableLockNone, 0, nil
+	}
+	return record.lockType, record.lockLease, nil
+}
+
+func (r *mockStateRemoteData) LockForRead(tid int64, now, ts uint64) (bool, error) {
+	record, ok := r.data[tid]
+	if !ok {
+		record = &stateRecord{
+			lockLease:    ts,
+			oldReadLease: ts,
+			lockType:     CachedTableLockRead,
+		}
+		r.data[tid] = record
+		return true, nil
+	}
+	switch record.lockType {
+	case CachedTableLockNone:
+		// Add the read lock
+		record.lockType = CachedTableLockRead
+		record.lockLease = ts
+		record.oldReadLease = ts
+		return true, nil
+	case CachedTableLockRead:
+		// Renew lease for this case.
+		if record.lockLease < ts {
+			record.lockLease = ts
+			record.oldReadLease = ts
+			return true, nil
+		}
+		// Already read locked.
+		return true, nil
+	case CachedTableLockWrite, CachedTableLockIntend:
+		if now > record.lockLease {
+			// Outdated...clear orphan lock
+			record.lockType = CachedTableLockRead
+			record.lockLease = ts
+			record.oldReadLease = ts
+			return true, nil
+		}
+		return false, nil
+	}
+	return false, errors.New("unknown lock type")
+}
+
+func (r *mockStateRemoteData) LockForWrite(tid int64, now, ts uint64) (uint64, error) {
+	record, ok := r.data[tid]
+	if !ok {
+		record.lockType = CachedTableLockWrite
+		record.lockLease = ts
+		record.oldReadLease = ts
+		return 0, nil
+	}
+
+	switch record.lockType {
+	case CachedTableLockNone:
+		record.lockType = CachedTableLockWrite
+		record.lockLease = ts
+		return 0, nil
+	case CachedTableLockRead:
+		if now > record.lockLease {
+			// Outdated, clear orphan lock and add write lock directly.
+			record.lockType = CachedTableLockWrite
+			record.lockLease = ts
+			return 0, nil
+		}
+
+		// Change state to intend, prevent renew lease operation.
+		oldLease := record.lockLease
+		record.lockType = CachedTableLockIntend
+		record.lockLease = leaseFromTS(ts)
+		record.oldReadLease = oldLease
+		return oldLease, nil
+	case CachedTableLockWrite:
+		if ts > record.lockLease {
+			record.lockLease = ts
+		}
+	case CachedTableLockIntend:
+		// Add the write lock.
+		if now > record.oldReadLease {
+			record.lockType = CachedTableLockWrite
+			record.lockLease = ts
+		}
```

Comment:
else branch should return `record.oldReadLease` to notify the client to wait!

Reply:
okay .i have done it